Исправлено: запускается из консоли

Uliana-Afonina opened the pull request.

Uliana-Afonina committed:
Исправлено: запускается из консоли

MrLukas007 commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+        Logger.info("Открыт браузер");
+
+
+        searchPage.checkBox223FL.waitUntil(Condition.visible, 10000, 50).setSelected(true);//чек 223-ФЗ
```
Сетеры, гетеры, нажимания кнопок, ожидания и другие методы связанные связанные с конкретной страницей должны быть вынесены в Page Object, в данном случае в SearchPage. Page Object это не только поля (как тут реализовано), но и методы для работы с элементами страницы. 
Замечание относится не только данному месту, но и к тесту в целом.

MrLukas007 commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+
+        searchPage.checkBox223FL.waitUntil(Condition.visible, 10000, 50).setSelected(true);//чек 223-ФЗ
+        Logger.info("Чек-бокс '223-ФЗ' выбран");
+        searchPage.checkBoxCP.waitUntil(Condition.visible, 10000, 50).setSelected(true);//чек Коммерческая закупка
```
Что такое 10000? 50? Данные значения, тем более они используются во многих местах, должны быть вынесены в отдельные переменные и корректно названы.

MrLukas007 commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+        }
+
+        searchPage.buttonSearch.waitUntil(Condition.visible, 10000, 50).click();//нажатие кнопки Поиск
+        Logger.info("Нажата кнопка 'Поиск'");
```
Логер тоже возможно вынести в Page Object.

MrLukas007 commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+                    if (trade.parent().find(searchPage.price).text().matches("(.*)EUR(.*)")) {
+                        currentSum = Double.parseDouble(
+                                trade.parent().find(searchPage.price).text()
+                                        .replace(" ", "")
```
Модификацию из текста в поле в double нужно делать делать в отдельном методе. Который по хорошему должен не принадлежать ни к тесту ни к Page Object. Возможно создать новый package (назвать его к примеру helpers) и создать там класс Util и в нём создать нужный метод для преобразования. Метод должен работать для любой валюты доступной в тесте.

MrLukas007 replied:
Возможно даже упростить логику получения денег.
1. Брать подстроку с начала строки до пробела.
2. У получившейся строки делать .replace(",", ".")

MrLukas007 commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+                    //переводим строку с ценой в числовой формат, откидываем нечисленные символы
+                    //проверяем, в какой валюте указана цена. Переводим в рубли, если есть необходимость
+                    //курс евро и доллара подтягивается из файла Data.ini
+                    if (trade.parent().find(searchPage.price).text().matches("(.*)EUR(.*)")) {
```
Лучше использовать не if else, a switch case.

dekkir replied:
длинные последовательности цепочек if-else следует избегать путем:
1) использования switch
2) изменения логики кода так, чтоб такие цепочки перестают быть нужными

MrLukas007 commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+                    //переводим числа из экспоненциального формата в "читабельный"
+                    BigDecimal bd = new BigDecimal(currentSum).setScale(2, ROUND_DOWN);
+
+                    // System.out.println("Trade #" + trade.text() + " Sum:" + bd);
```
Думаю это можно удалить.

MrLukas007 commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+
+        double usd = 0.0;
+        double eur = 0.0;
+        try {
```
Если файла нет, то тест упадёт. Возможно стоит проверять файл в начале или даже перед запуском теста или тестов. Посмотреть в сторону @BeforeAll @BeforeEach у Junit.
И делать проверку только наличия файла, получать значения нужно уже в тесте.

sevnight commented on pom.xml:
```diff
@@ -0,0 +1,82 @@
+        <dependency>
+            <groupId>com.codeborne</groupId>
+            <artifactId>selenide</artifactId>
+            <version>5.13.0</version>
```
Версии удобно выносить чтобы было проще их контролировать.
![image](https://user-images.githubusercontent.com/29500839/89014792-bbb67d00-d31e-11ea-95a8-9d77ebc731d8.png)

sevnight reviewed (changes requested):
Нужно привести проект к корректной структуре.
![image](https://user-images.githubusercontent.com/29500839/89015132-3f706980-d31f-11ea-9342-854550ddb46a.png)

sevnight commented on src/test/java/rts/pages/SearchPage.java:
```diff
@@ -0,0 +1,35 @@
+    //поле ввода даты Конца извещения
+    public SelenideElement dateTo = $("#BaseMainContent_MainContent_txtPublicationDate_txtDateTo");
+    //кнопка, перелистывающая таблицу
+    public String nextPage = "#next_t_BaseMainContent_MainContent_jqgTrade_toppager";
```
Посмотри как в Java записываются строковые константы.

dekkir commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+        Logger.info("Чек-бокс 'Коммерческая закупка' выбран");
+        searchPage.startPrice.waitUntil(Condition.visible, 10000, 50).setValue("0");//установка начальной цены
+        Logger.info("Установлена начальная цена от 0");
+        //проверка исключения (вдруг файл не существует)
```
Кажется, комментарий должен быть не здесь, а ниже

dekkir commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+        Logger.info("Нажата кнопка 'Поиск'");
+        //ждем, пока прогрузится таблица (должно исчезнуть окно Загрузка...)
+        Logger.info("Ожидаю, пока прогрузится таблица");
+        if ($(searchPage.loader).isDisplayed()) {
```
думаю, здесь if лишний, поскольку waitUntil выполняет ту же функцию: будет ждать, пока окно с загрузкой будет удовлетворять условию not(visible)

MrLukas007 commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+                        currentSum = currentSum * eur;
+                        Logger.info("Цена #" + trade.text() + " была указана в EUR. Пересчитана в руб.");
+
+                    } else if (trade.parent().find(searchPage.price).text().matches("(.*)USD(.*)")) {
```
Достаточно сложная логика находится в if , так быть не должно. Возможно после вынесения методов в Page object логика станет проще.

MrLukas007 replied:
Не уверен что логика которая находится внутри matches правильная.
matches("(.*)USD(.*)") - означает что будет что-то до и после USD, но глядя на страницу после USD ничего нет.
Возможно этого будет достаточно matches("(.*)USD)
Тоже самое и с EUR. C руб немного по другому там будет не "руб", а "руб.", тогда и с руб будет работать. Так же возможно просто использование метода contains. https://www.w3schools.com/java/ref_string_contains.asp

dekkir commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+        //до тех пор, пока кнопка перехода на следующую страницу таблицы кликабельная,
+        //считаем количество и сумму лотов
+        do {
+            if ($(searchPage.loader).isDisplayed()) {
```
думаю, здесь if лишний, поскольку waitUntil выполняет ту же функцию: будет ждать, пока окно с загрузкой будет удовлетворять условию not(visible)

dekkir commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+                $(searchPage.loader).waitUntil(not(Condition.visible), 10000, 50);
+            }
+            //поиск по коллекции. Выбираем те строки, где присутствует номер ЕИС
+            ElementsCollection trades = Selenide.$$(searchPage.oosNumber);
```
Эту строчку и следующую можно объединить в одну 
ElementsCollection trades = Selenide.$$(searchPage.oosNumber).filterBy(not(Condition.empty));

dekkir commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+                                        .replace(",", ".")
+                                        .replace("EUR", "")
+                        );
+                        currentSum = currentSum * eur;
```
такую запись можно сокращенно писать:
currentSum *= eur

dekkir commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+                    //курс евро и доллара подтягивается из файла Data.ini
+                    if (trade.parent().find(searchPage.price).text().matches("(.*)EUR(.*)")) {
+                        currentSum = Double.parseDouble(
+                                trade.parent().find(searchPage.price).text()
```
повторяющийся код:
```
trade.parent().find(searchPage.price).text()
.replace(" ", "")
 .replace(",", ".")
.replace("EUR", "")
```
следует выносить в отдельный метод и вызывать, где он необходим

dekkir replied:
В данном случае метод будет параметризированным, и принимать на вход тип валюты

MrLukas007 commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+
+    @Test
+    public void countSums() throws IOException {
+        Selenide.open("https://223.rts-tender.ru/supplier/auction/trade/search.aspx");
```
Урл стоит вынести в переменную.

MrLukas007 commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+        }
+
+        double sum = 0.0;
+        int kol = 0;
```
Что такое kol?

MrLukas007 commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,158 @@
+        String text = "Количество лотов: " + kol + "\nСумма лотов: " + allSum + " руб.";
+        Logger.info("Количество лотов: " + kol + "\nСумма лотов: " + allSum + " руб.");
+        //создаем новый экземпляр файла по заданному пути
+        File file = new File("src/test/java/rts/files/SumKol.txt");
```
Что за название SumKol?

Uliana-Afonina committed:
Модификацию из текста в поле в double сделала в package "helpers" в классе Util (создала нужный метод для преобразования). Метод работает для валют из теста.
Отказалась от структуры if else в пользу switch case.
Урл вынесла в переменную в SearchPage .
Переименовала переменную kol в count. Переменная отвечает за количество лотов.
Переименовала файл с результатами в SumCountLots.
Значения для ожидания и интервала опроса занесла в переменные wait и waitInterval соответственно.

Uliana-Afonina committed:
Переписала логику преобразования цены лота из строки в вещественное число и конвертацию валюты, а именно: создала метод, который преобразует цену лота из строки в число, в котором вызвала метод, конвертирующий валюту в случае, если это требуется.
Сделала проверку наличия файла Data.ini перед началом работы теста (использовала аннотацию @BeforeEach)

Uliana-Afonina committed:
Привела проект к корректной структуре.
В SearchPage вынесла методы для работы с элементами SearchTest.
В класс Util вынесла аннотацию @BeforeEach для проверки существования файла Data.ini. Запись результата в файл SumCountLots.txt оформила методом и тоже вынесла в Util.

Uliana-Afonina committed:
Привела проект к корректной структуре.
В SearchPage вынесла методы для работы с элементами SearchTest.
В класс Util вынесла аннотацию @BeforeEach для проверки существования файла Data.ini. Запись результата в файл SumCountLots.txt оформила методом и тоже вынесла в Util.

MrLukas007 commented on src/main/java/helpers/Util.java:
```diff
@@ -0,0 +1,93 @@
+    //В prOfLt передать priceOfLot
+    //В rate передать eur либо usd (выбор делается в методе convertStringToDouble()
+
+    private static Double convertCurrency(double prOfLt, double rate) {
```
Данный метод нужно удалить, т.к. он ничего не делает кроме умножения.

MrLukas007 commented on src/main/java/helpers/Util.java:
```diff
@@ -0,0 +1,93 @@
+// преобразуем строку в вещественное число
+        switch (priceText.substring(i + 1)) {
+            case "EUR":
+                priceOfLot = Double.parseDouble(priceText
```
Уже писалось ранее, но сделано не было. Нужно вынести в отдельный метод преобразование в из текста с валютой в дабл!

MrLukas007 commented on src/main/java/helpers/Util.java:
```diff
@@ -0,0 +1,93 @@
+                        .replace("EUR", ""));
+                priceOfLot = convertCurrency(priceOfLot, eur);
+                Logger.info("Цена #" + tradeText + " была указана в EUR. Пересчитана в руб.");
+                break;
```
Так же говорилось ранее про ненужный break, нужно делать достаточно return!

MrLukas007 commented on src/main/java/pages/SearchPage.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    //переключаем на следующую страницу таблицы
+    public void clickNextPage(String nxtPg) {
```
Зачем метод принимает какое-то строчное значение? Используейте поля класса для работы в методах.
Это касается всех Page object методов.

MrLukas007 commented on src/test/java/SearchTest.java:
```diff
@@ -0,0 +1,101 @@
+
+            for (SelenideElement trade : trades) {
+                //Выбираем те значения, статус которых не отменен
+                if (!trade.parent().find(searchPage.state).text().equals("Отменена")) {
```
Нужно сделать методы в Page Object которые будут возвращаться state или price или любое другое значение. А передавать в этом метод нужно trade, внутри которого и будет делаться поиск элемента и браться значение.

MrLukas007 commented on src/test/java/rts/pages/SearchPage.java:
```diff
@@ -0,0 +1,44 @@
+package rts.pages;
+
```
Зачем нужен данный файл, если создан новый в нужной дериктории?
Так же замечание ниже про строковые константы не было сделано.

Uliana-Afonina committed:
В Util переделала методы для преобразования цены лота из строки в вещественное число конвертации валюты.

В SearchPage:
 -вынесла в метод поиска элемента, отвечающего за цену лота;
 -объявила строковые константы как public static final;
 -использовала поля класса для работы в методах.

MrLukas007 commented on src/main/java/helpers/Util.java:
```diff
@@ -0,0 +1,84 @@
+    //В rate передать eur либо usd (выбор делается в методе convertCurrency()
+    public static Double convertStringToToDouble(String priceText, String rate) {
+
+        priceOfLot = Double.parseDouble(priceText
```
https://github.com/Uliana-Afonina/Solar_Lab_Test/pull/1#discussion_r463514750
Упростить

Uliana-Afonina replied:
Сделала. Но не считаю, что это упрощает. Добавилось лишнее действие. А так как пробелов в строке может быть больше одного, то нужно искать последний пробел и делать replace всех остальных пробелов.

MrLukas007 commented on src/main/java/helpers/Util.java:
```diff
@@ -0,0 +1,84 @@
+    //в priceText передать String textCurrency = trade.parent().find(searchPage.price).text();
+    //В rate передать eur либо usd (выбор делается в методе convertCurrency()
+    public static Double convertStringToToDouble(String priceText, String rate) {
+
```
Убрать лишнюю пустую строку

MrLukas007 commented on src/main/java/helpers/Util.java:
```diff
@@ -0,0 +1,84 @@
+    //В rate передать eur либо usd (выбор делается в методе convertCurrency()
+    public static Double convertStringToToDouble(String priceText, String rate) {
+
+        priceOfLot = Double.parseDouble(priceText
```
Не нужно записывать в переменную значение а потом её возвращать. Нужно сразу делать return и потом то что нужно вернуть. Например:
```
int a;
a = b*c;
return a;
```
Этот вариант плохой.
`return b*c;` - этот хороший.

MrLukas007 commented on src/main/java/helpers/Util.java:
```diff
@@ -0,0 +1,84 @@
+
+public class Util {
+    //инициализация переменной для хранения значения текущей стоимости
+    public static double priceOfLot = 0.00;
```
Эта переменная не должна быть полем данного класса, а должна быть переменной в конкретном методе.

MrLukas007 commented on src/main/java/helpers/Util.java:
```diff
@@ -0,0 +1,84 @@
+            case "EUR":
+                priceOfLot = convertStringToToDouble(priceText, "EUR") * eur;
+                Logger.info("Цена #" + tradeText + " была указана в EUR. Пересчитана в руб.");
+                return priceOfLot;
```
Тоже самое замечание про return. Тут нужно возвращать сразу результат, а не переменную. Исправить во всех местах в данном методе.

MrLukas007 commented on src/main/java/pages/SearchPage.java:
```diff
@@ -0,0 +1,116 @@
+    public String textCurrency;
+
+    //ищем элемент, отвечающий за цену лота
+    public String findPrice(SelenideElement trade) {
```
метод должен называться getPrice, который будет только возвращаться цену закупки.
т.е. условие - должно уехать в тест, а в PO(Page Object) остаться просто возвращение значения.

MrLukas007 commented on src/test/java/rts/pages/SearchPage.java:
```diff
@@ -0,0 +1,44 @@
+package rts.pages;
```
Еще раз напишу, что данный файл нужно удалить!

MrLukas007 commented on src/main/java/pages/SearchPage.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    //переключаем на следующую страницу таблицы
+    public void clickNextPage(String NEXT_PAGE) {
```
Не сделано замечание https://github.com/Uliana-Afonina/Solar_Lab_Test/pull/1#discussion_r464219600 !!!

MrLukas007 commented on src/main/java/pages/SearchPage.java:
```diff
@@ -0,0 +1,116 @@
+
+    //выбор чек-боксов
+    public void setCheckBox(SelenideElement checkBox) {
+
```
Удалить лишние пустые строки в начале и в конце в этом и других методах. Пустые строки можно добавлять в методы для визуального разделения и читабельности, в методах из 2-3-4 строк это безсмысленно.

MrLukas007 commented on src/main/java/pages/SearchPage.java:
```diff
@@ -0,0 +1,116 @@
+    public String findPrice(SelenideElement trade) {
+        //Выбираем те значения, статус которых не отменен
+        if (!trade.parent().find(STATE).text().equals("Отменена")) {
+            textCurrency = trade.parent().find(PRICE).text();
```
Зачем делать trade.parent().find(PRICE).text()?
Можно просто найти элемент внутри элемента, и взять из него текст.
`trade.$x("PRICE_LOCATOR").getText();`

Uliana-Afonina replied:
Сделала. Но использовать parent() все равно пришлось, так как Collection по номеру в ЕИС. А со столбцом цены у них общий предок.

MrLukas007 commented on src/main/java/pages/SearchPage.java:
```diff
@@ -0,0 +1,116 @@
+    //ячейка с таблицы с номером ЕИС
+    public final String OOS_NUMBER = "[aria-describedby=\"BaseMainContent_MainContent_jqgTrade_OosNumber\"]";
+    //ячейка таблицы со статусом заявки
+    public final String STATE = "[aria-describedby=\"BaseMainContent_MainContent_jqgTrade_LotStateString\"]";
```
Стоковые переменный должны быть переделаны в локаторы подобного вида
.//*[@aria-describedby='BaseMainContent_MainContent_jqgTrade_StartPrice']

MrLukas007 commented on src/main/java/pages/SearchPage.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    //инициализируем строковую переменную для хранения цены лота
+    public String textCurrency;
```
Это нужно срочно удалить! В PO должны быть гетеры (методы которые возвращают определенные значения)! В вашем гетере нужно просто возвращать значение.

Uliana-Afonina committed:
попытка удалить с гитхаб rts/files и rts/SearchPage

Uliana-Afonina committed:
Метод convertStringToToDouble переписала, изменив логику на поиск последнего пробела.
Заменила break на return. Через return возвращаю сразу результат, а не переменные.
В методах Page object использовала поля класса.
Убрала лишние пустые строки.
Переделала строковые константы OOS_NUMBER, STATE и PRICE.

Uliana-Afonina committed:
Переделала поля класса SearchPage LOADER и NEXT_PAGE из строковых констант в SelenideElement.
Написала для каждого чек-бокса свой метод.
setCheckBox223FL(),setCheckBoxCP(),clickNextPage(), waitLoading() используют поля класса SearchPage.

Uliana-Afonina committed:
Переделала поля класса SearchPage LOADER и NEXT_PAGE из строковых констант в SelenideElement.
Написала для каждого чек-бокса свой метод.
setCheckBox223FL(),setCheckBoxCP(),clickNextPage(), waitLoading() используют поля класса SearchPage.

Uliana-Afonina committed:
Переделала поля класса SearchPage LOADER и NEXT_PAGE из строковых констант в SelenideElement.
Написала для каждого чек-бокса свой метод.
setCheckBox223FL(),setCheckBoxCP(),clickNextPage(), waitLoading() используют поля класса SearchPage.
